go/protocols/materialize/transactor.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	var rxRequest = pm.Request{Open: &open}
+	var txResponse, err = WriteOpened(stream, &opened)
+	if err != nil {
+		return err
+	}
+
+	var (
+		// awaitErr is the last await() result,
+		// and is readable upon its close of its parameter `awaitDoneCh`.
+		awaitErr error
+		// loadErr is the last loadAll() result,
+		// and is readable upon its close of its parameter `loadDoneCh`.
+		loadErr error
+	)
+
+	// await is a closure which awaits the completion of a previously
+	// started commit, and then writes Acknowledged to the runtime.
+	// It has an exclusive ability to write to `stream` until it returns.
+	var await = func(
+		round int,
+		ourCommitOp OpFuture, // Resolves when the prior commit completes.
+		awaitDoneCh chan<- struct{}, // To be closed upon return.
+		loadDoneCh <-chan struct{}, // Signaled when load() has completed.
+	) (__out error) {
+
+		defer func() {
+			logrus.WithFields(logrus.Fields{
+				"round": round,
+				"error": __out,
+			}).Debug("await commit finished")
+
+			awaitErr = __out
+			close(awaitDoneCh)
+		}()
+
+		var ackState *pf.ConnectorState = nil
+
+		// Wait for commit to complete, with cancellation checks.
+		select {
+		case ackState = <-ourCommitOp.Done():
+			if err := ourCommitOp.Err(); err != nil {
+				return err
+			}
+		case <-loadDoneCh:
+			// load() must have error'd, as it otherwise cannot
+			// complete until we send Acknowledged.
+			return nil
+		}
+
+		return WriteAcknowledged(stream, ackState, &txResponse)
+	}
+
+	// load is a closure for async execution of Transactor.Load.
+	var load = func(
+		round int,
+		it *LoadIterator,
+		loadDoneCh chan<- struct{}, // To be closed upon return.
+	) (__out error) {
+
+		var loaded int
+		defer func() {
+			logrus.WithFields(logrus.Fields{
+				"round":  round,
+				"total":  it.Total,
+				"loaded": loaded,
+				"error":  __out,
+			}).Debug("load finished")
+
+			loadErr = __out
+			close(loadDoneCh)
+		}()
+
+		var err = transactor.Load(it, func(binding int, doc json.RawMessage) error {
+			if it.err != nil {
+				panic(fmt.Sprintf("loaded called without first checking LoadIterator.Err(): %v", it.err))
+			} else if it.awaitDoneCh != nil {
+				panic("loaded called without first calling LoadIterator.WaitForAcknowledged()")
+			} else if awaitErr != nil {
+				// We cannot write a Loaded response if await() failed, as it would
+				// be an out-of-order response (a protocol violation). Bail out.
+				return context.Canceled
+			}
+
+			loaded++
+			return WriteLoaded(stream, &txResponse, binding, doc)
+		})
+
+		if it.awaitDoneCh == nil && awaitErr != nil {
+			return nil // Cancelled by await() error.
+		} else if it.err != nil {
+			// Prefer the iterator's error over `err` as it's earlier in the chain
+			// of dependency and is likely causal of (or equal to) `err`.
+			return it.err
+		}
+		return err
+	}
+
+	// ourCommitOp is a future for the last async startCommit().
+	var ourCommitOp OpFuture = FinishedOperation(nil)
```

Comment:
Agreed. It's poorly defined, right now, where one is supposed to do this. As-implemented (because Transactor doesn't address this) you've got an initial apply happening upon open within the details of the Snowflake implementation. But, as noted, that apply doesn't get to update the connector state. It would be better if Transactor could drive this post-open apply behavior.

One issue is the definition of StartCommitFunc. I'm thinking we need to tease apart StartCommit and Acknowlege as separate callbacks which are driven by Transactor. Trying this on:

```
// StartCommitFunc begins to commit a stored transaction.
// Upon its return a commit operation may still be running in the background,
// and the returned OpFuture must resolve with its completion.
type StartCommitFunc = func(context.Context, *pc.Checkpoint) (*pf.ConnectorState, OpFuture)

interface Transactor {
  Load()
  Store() (StartCommitFunc, error)
  
  // Acknowledge the commit of a completed transaction.
  // Acknowledge is run after both a) request.Acknowledge has been received from the runtime,
  // and also b) after an OpFuture returned by StartCommit has resolved.
  // It takes a ConnectorState which may either be a recovered ConnectorState upon startup,
  // or a ConnectorState which was just returned by a preceding StartCommitFunc.
  // It returns an optional ConnectorState update which will be applied in a best-effort fashion
  // upon its successful completion.
  //
  // Acknowledge may perform long-running, idempotent operations such as merging staged
  // updates into a base table. Upon its succesful return, response.Acknowledged is sent to
  // the runtime, allowing the next pipelined transaction to begin to close.  
  Acknowledge(*pf.ConnectorState) (*pf.ConnectorState, error)
}
```

* Note that we no longer send response.Acknowledged when the StartCommitFunc future resolves. That happens only after Acknowledge() completes.

* RunTransactions could then be responsible for driving Acknowledge() on startup, using the recovered connector state, and also for driving it during transactions.

* StartCommit _could_ be a callback of Transactor which synchronously COMMITs a stored transaction. However, the current callback behavior of StartCommit was done to make it easier to pass around opened transaction variables, which would otherwise need to be preserved on the Transactor-implementing struct, which gets weird.

* I recognize this is back-tracking on the conversation above, that ConnectorStates returned by Acknowledge shouldn't depend on the actual execution of the apply. Acknowledge could return an OpFuture itself to make this relationship more clear, but that makes the interface unnecessarily awkward IMO.

Reply:
Thinking more about this, if we are moving the Acknowledge phase to a method of its own, why would we still allow an `OpFuture` from `StartCommit`, instead of only from `transactor.Acknowledge`? Any async operation will need to wait for the runtime ack to come before attempting to actually commit the data. Previously this was done by having the async operation of `StartCommit` wait for `runtimeAckCh`, but now that `transactor.Acknowledge` runs after the async request from runtime has come, it makes sense to me that `StartCommit` no longer should return an async operation. It should only return the checkpoint for the staged operations, and in `transactor.Acknowledge` the actual process of applying these staged operations should take place.

Reply:
We still need StartCommit to return an OpFuture because the sequence is:

* Runtime sends a bunch of request.Store, followed by a request.StartCommit
* RunTransactions runs the StartCommitFunc returned by Store, and on it's return it _immediately_ sends StartedCommit.
  * We don't want to block here, because it lets the runtime write its recovery log checkpoint as quickly as possible and start work on the next transaction.
* RunTransactions looks for a request.Acknowledge intermixed with request.Load of the next transaction. It also waits for the OpFuture returned by StartCommitFunc to resolve. Once both have happened, then it calls impl.Acknowledge

Implementations like materialize-postgres would have meaningful StartCommitFunc implementations that actually COMMIT, and then a no-op Acknowledge. We do want to send response.StartedCommit _before_ the started COMMIT actually completes.

> One weird part though is that in order for us to pass the latest state of the connector to transactor.Acknowledge, we need to know the resulting checkpoint after StartCommitted has been processed by the runtime, that is, we need request.Acknowledge to include the resulting state of the connector. 

Yea... 🤔  What happens if we introduce a `State` interface, that RunTransactions owns but passes into Transactor callbacks ? The State interface would know how to marshal its incremental changes since the last time it was asked to do so. Something like:

```
interface State {
   // Replace this State with the recovered checkpoint provided by the Flow runtime.
   // Unmarshal is called only on transactor startup, on a State returned by Transactor.NewState.
   Unmarshal(json.RawMessage) error

   // Marshal a ConnectorState which reflects the incremental changes of this connector
   // state since the last call to MarshalChanges. 
   MarshalChanges() (*pf.ConnectorState, error)
}

interface Transactor {
  NewState() State
}
```

Reply:
Or perhaps I'm over thinking this: Transactor could instead hold the current state as an inner field of its implementation, and have a function to rehydrated its inner state from a persisted checkpoint:

```
interface Transactor {
  // Called on startup if there is a persisted state.
  UnmarshalState(json.RawMessage) error

  Load() // unchanged
  Store() (StartCommitFunc, error) // unchanged.

  // Runs synchronously to apply the current state, which is populated by
  // RestoreState _or_ a prior StartCommitFunc.
  Acknowledge() (*pf.ConnectorState, error)
}

```

Reply:
@jgraettinger I don't see how this interface change addresses the challenge. The challenge is that I want there to be consistency between the _kind of_ state `Acknowledge` receives. I don't want it to sometimes receive a recovered checkpoint, and sometimes an incremental change on top of the last checkpoint. So ideally I would want `Acknowledge` to receive `json.RawMessage` as state, not `pf.ConnectorState`, since `pf.ConnectorState` is an operation to update the last state, not the final state itself. When the connector boots up first, we would need to construct an artificial `pf.ConnectorState` from the recovered checkpoint, which smells to me.

I think to address this we don't need to involve the transactor. If the idea is to replicate the logic of merging state updates in our library, then we can do that in `RunTransactions` without involving the Transactor to implement another method. So in `RunTransactions` we can take the initial checkpoint from the `open.Request` (a `json.RawMessage`), and then for every `pf.ConnectorState` we receive from `startCommit` and `Acknowledge`, if they are merge patches, we apply the merge patch to the last known state to end up with a new `json.RawMessage`, and use that for the next call to `Acknowledge`.

This would mean replicating the state reduction logic though, which is a bit gross. It doesn't seem very likely that we want to change this logic in flow runtime, but having this logic in two places creates the possibility of breaking something if we do.

Reply:
I've implemented what I have in mind in my last commit, let me know what you think of this: https://github.com/estuary/connectors/pull/1117/commits/f6e85f9a928a1fdd184dd6669b63822212807e60